README.md
# kar-traffic-fines 

[![Github CI](https://github.com/nikhilnagaraju/kar-traffic-fines/workflows/Node.js%20CI/badge.svg)](https://github.com/nikhilnagaraju/kar-traffic-fines/actions?query=workflow%3A%22Node.js+CI%22)
[![dependencies Status](https://david-dm.org/nikhilnagaraju/kar-traffic-fines/status.svg)](https://david-dm.org/nikhilnagaraju/kar-traffic-fines)
[![devDependencies Status](https://david-dm.org/nikhilnagaraju/kar-traffic-fines/dev-status.svg)](https://david-dm.org/nikhilnagaraju/kar-traffic-fines?type=dev)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-blue.svg)](https://conventionalcommits.org)

Simple **JSON** API to fetch traffic fines for any KA registered vehicle. Can consider this as a simple proxy without CORS Blocking.
Hosted on [Heroku](https://heroku.com).

### Usage

|Title|Get Traffic Fines|
|----|----|
| Endpoint | `https://kar-traffic-fines.herokuapp.com` |
| Path | `/fines` |
| Method | `GET` |
| URL params | Required: `registration`=`KAXX(AZ)XXXX` |
| Error response <br> *Invalid Registration, Missing Registration param* | status code: `400`  <br> content: ```{status: 'error',message: 'A valid 'registration' param in 'KAXX(AZ)XXXX' format is required.'}```  |
| Example request | `https://kar-traffic-fines.herokuapp.com/fines?registration=KA02AB0123` |


#### Why is this API useful?
* No CORS Header blocking! Can be used on any client app, sites.
* JSON API to fetch/check fines. While you may find other APIs which gives a data in not-so-usable format.



## Development

Install dependancies
* `npm install` or `yarn`

Start dev server with file change watch
* `npm run serve` or `yarn serve`

Run linter
* `npm run lint` or `yarn lint`

Run Tests
* `npm run test` or `yarn test`

Run coverage
* `npm run coverage` or `yarn coverage`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/nikhilnagaraju/kar-traffic-fines/compare/v1.0.0...v1.0.1) (2020-04-29)


### Bug Fixes

* Removed unused try/catch ([d86ed44](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/d86ed440b95076887e4243dd0c5801ebeade7c5a))

## 1.0.0 (2020-04-29)


### Features

* base implementation ([88e0668](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/88e0668216ed151db893185b7b765be88f19b5a3))
* eslint added ([c52768c](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/c52768c9b6a771f4d057c2a43d6a6d9284fbf6d1))
* Header support for Allowing any origin ([aef7d28](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/aef7d28593e4faf9a3218e1b07e8801fe65d089d))
* Procfile for heroku deployment ([68e874a](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/68e874a5b0ba272ecdd0b47b8940248d740338f7))


### Bug Fixes

* **script:** Versioning with changelog generation ([ee0a46e](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/ee0a46e0e53d0ac5d1055be86062cc697129f626))
* basic checks and error handling ([2e96a15](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/2e96a153c6c7b0255bdf0a134067b21cb52a786d))
* eslint format ([d2fda4f](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/d2fda4f8d4ceac14057664133b638a17bab86675))
* export app module ([ebbb06b](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/ebbb06bcb0aff4e73f385ff5e2ac499c8b848b8a))
* Test runner configuration with ava, supertest ([714bf0f](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/714bf0fd5d5c1cdf843eab7958e807da55b7152a))
* Valid Registration regex ([8fdb5cf](https://github.com/nikhilnagaraju/kar-traffic-fines/commit/8fdb5cf1d1ab4b4f1d4a85ebfea298289043d5a4))


test/test.spec.js
import test from'ava';
import{ agent }from'supertest';
import app from'../app';
const appServer = agent(app);

const registrationPattern = RegExp(/^ka(?!00)([0-6][0-9]|70)[a-z]{0,2}[0-9]{4}$/i);

test('[Success] Server status check', async (t) => {
   const res = await appServer.get('/').expect(200);
   const expectedResponse = {
      status: 'ok',
      message: 'Why say hello world!, when you can say HELLO UNIVERSE!'
   };
   t.deepEqual(res.body, expectedResponse);
});

test('[Sucess] Valid Registration', (t) => {
   t.regex('KA021234', registrationPattern, 'Reg. No without category');
   t.regex('KA02A1234', registrationPattern, 'Reg. No with single alphabet category');
   t.regex('KA02AB1234', registrationPattern, 'Reg. No with category');
   t.regex('KA02AB0034', registrationPattern, 'Reg. No with prepended 0\'s');
   t.regex('ka02ab0034', registrationPattern, 'Lowercase Reg. No');
});

test('[Fail] Invalid Registration', (t) => {
   t.notRegex('KA1234', registrationPattern, 'No RTO');
   t.notRegex('KA001234', registrationPattern, 'Invalid 00 RTO');
   t.notRegex('KA71AB1234', registrationPattern, 'Invalid RTO');
   t.notRegex('KA02AB12', registrationPattern, 'Invalid number');
   t.notRegex('GJ02A1234', registrationPattern, 'Non Karnataka Reg. No');
});

test('[Success] Valid Karnataka One response', async (t) => {
   const res = await appServer.get('/fines').query({ registration: 'KA02AB1234' }).expect(200);
   t.truthy('PoliceFineDetailsList' in res.body);
   t.truthy('TotalFineAmount' in res.body);
   t.truthy('Response' in res.body);
});

test('[Fail] Missing registration param', async (t) => {
   const res = await appServer.get('/fines').expect(400);
   const expectedResponse = {
      status: 'error',
      message: 'A valid \'registration\' param in \'KAXX(AZ)XXXX\' format is required.'
   };
   t.deepEqual(res.body, expectedResponse);
});

test('[Fail] Invalid registration param', async (t) => {
   const res = await appServer.get('/fines').query({ registration: 'KA1234' }).expect(400);
   const expectedResponse = {
      status: 'error',
      message: 'A valid \'registration\' param in \'KAXX(AZ)XXXX\' format is required.'
   };
   t.deepEqual(res.body, expectedResponse);
});

test('[Fail] Invalid endpoints', async (t) => {
   const res = await appServer.get('/random/endpoint').expect(400);
   const expectedResponse = {
      status: 'error',
      message: 'use \'/fines\' endpoint with \'registration\' query param in \'KAXX(AZ)XXXX\' format to check fines.'
   };
   t.deepEqual(res.body, expectedResponse);
});


app.js
const express = require('express');
const bodyParser = require('body-parser');
const logger = require('morgan');
const responseTime = require('response-time');
const appRoutes = require('./routes');

const app = express();

app.set('port', process.env.PORT || 8989);

app.use(bodyParser.urlencoded({ extended: false }));
app.use(logger('common'));
app.use(responseTime());
app.use(appRoutes);
app.use((req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET');
  res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type');
  res.setHeader('Access-Control-Allow-Credentials', true);
  next();
});

app.listen(app.get('port'), () => {
  console.log(`server is listening on ${app.get('port')}`);
});

module.exports = app;


routes.js
const express = require('express');
const router = express.Router();
const axios = require('axios');
const trafficFinesEndpoint = 'https://www.karnatakaone.gov.in/PoliceCollectionOfFine/FineDetails';
const regPattern = RegExp(/^ka(?!00)([0-6][0-9]|70)[a-z]{0,2}[0-9]{4}$/i);

router.get('/', (req, res) => {
  res.send({
    status: 'ok',
    message: 'Why say hello world!, when you can say HELLO UNIVERSE!'
  });
});

router.get('/fines', (req, res) => {
  const registration = req.query.registration;

  if(!registration || !regPattern.test(registration)){
    return res.status(400).send({
      status: 'error',
      message: 'A valid \'registration\' param in \'KAXX(AZ)XXXX\' format is required.'
    });
  }

  const options = {
    params: {
      SearchBy: 'REGNO',
      ServiceCode: 'BPS',
      SearchValue: registration
    }
  };
  return axios.post(trafficFinesEndpoint, null, options).then(extResponse => {
    res.send(extResponse.data);
  }).catch(err => {
    console.error(err);
    res.status(400).send(err);
  });
});

router.get('**', (req, res) => {
  res.status(400).send({
    status: 'error',
    message: 'use \'/fines\' endpoint with \'registration\' query param in \'KAXX(AZ)XXXX\' format to check fines.'
  });
});

module.exports = router;
